internal/configfsm/fsm.go

```diff
@@ -0,0 +1,278 @@
+	fsm.Configure(Missing).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + Missing)
+			fsmHelper.ResetNonCompliantAt()
+			if fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(MissingToClusterNotReady, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(MissingToClusterNotReady, ClusterNotReady).
+		PermitReentry(MissingToMissing)
+	fsm.Configure(ClusterNotReady).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + ClusterNotReady)
+			if !fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(ClusterNotReadyToMissing, fsmHelper)
+			}
+			if fsmHelper.IsClusterReady() ||
+				fsmHelper.IsTimedOut() {
+				return fsm.Fire(ClusterNotReadyToInProgress, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(ClusterNotReadyToMissing, Missing).
+		Permit(ClusterNotReadyToInProgress, InProgress).
+		PermitReentry(ClusterNotReadyToClusterNotReady)
+	fsm.Configure(InProgress).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + InProgress)
+			if fsmHelper.IsNonCompliantAtZero() ||
+				fsmHelper.IsAllPoliciesCompliant() ||
+				(!fsmHelper.IsNonCompliantPolicyInEnforce() && !fsmHelper.IsAllPoliciesCompliant()) {
+				fsmHelper.SetNonCompliantAtNow()
+			}
+			if !fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(InProgressToClusterNotReady, fsmHelper)
+			}
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(InProgressToTimedOut, fsmHelper)
+			}
+			if !fsmHelper.IsClusterReady() {
+				return fsm.Fire(InProgressToClusterNotReady, fsmHelper)
+			}
+			if !fsmHelper.IsNonCompliantPolicyInEnforce() && !fsmHelper.IsAllPoliciesCompliant() { // checks OutOfDate condition
+				return fsm.Fire(InProgressToOutOfDate, fsmHelper)
+			}
+			if fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(InProgressToCompleted, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(InProgressToClusterNotReady, ClusterNotReady).
+		Permit(InProgressToTimedOut, TimedOut).
+		Permit(InProgressToOutOfDate, OutOfDate).
+		Permit(InProgressToCompleted, Completed).
+		PermitReentry(InProgressToInProgress)
+	fsm.Configure(OutOfDate).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + OutOfDate)
+			fsmHelper.ResetNonCompliantAt()
+			if !fsmHelper.ArePoliciesMatched() ||
+				!fsmHelper.IsClusterReady() ||
+				fsmHelper.IsNonCompliantPolicyInEnforce() ||
+				fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(OutOfDateToInProgress, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(OutOfDateToInProgress, InProgress).
+		PermitReentry(OutOfDateToOutOfDate)
+	fsm.Configure(Completed).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering " + Completed)
+			fsmHelper.ResetNonCompliantAt()
+
+			if !fsmHelper.ArePoliciesMatched() ||
```

Comment:
It looks here like we cannot get to OutOfDate from Completed :thinking:

Reply:
Fixed! also added a unit test for it